posts/inside-rust/2019-11-23-const-if-match.md

```diff
@@ -0,0 +1,151 @@
+---
+layout: post
+title: "`if` and `match` in constants on nightly rust"
+author: Dylan MacKenzie
+---
+
+**TLDR; `if` and `match` are now usable in constants on the latest nightly.** 
+
+As a result, you can now write code like the following and have it execute at
+compile-time:
+
+```rust
+static PLATFORM: &str = if cfg!(unix) {
+    "unix"
+} else if cfg!(windows) {
+    "windows"
+} else {
+    "other"
+};
+
+const fn gcd(a: u32, b: u32) -> u32 {
+    match (a, b) {
+        (x, 0) | (0, x) => x,
+
+        (x, y) if x % 2 == 0 && y % 2 == 0 => 2*gcd(x/2, y/2),
+        (x, y) | (y, x) if x % 2 == 0 => gcd(x/2, y),
+
+        (x, y) if x < y => gcd((y-x)/2, x),
+        (x, y) => gcd((x-y)/2, y),
+    }
+}
+
+const _: () = assert!(std::mem::size_of::<usize>(), 8, "Only 64-bit platforms are supported");
+```
+
+## What exactly is going on here?
+
+The following expressions,
+- `match`
+- `if` and `if let`
+- `&&` and `||`
+
+can now appear in any of the following contexts,
+- `const fn` bodies
+- `const` and associated `const` initializers
+- `static` and `static mut` initializers
+- array initializers
+- const generics (EXPERIMENTAL)
+
+if `#![feature(const_if_match)]` is enabled for your crate.
+
+You may have noticed that the short-circuiting logic operators, `&&` and
+`||`, were already legal in a `const` or `static`. This was accomplished by
+translating them to their non-short-circuiting equivalents, `&` and `|`
+respectively. Enabling the feature gate will turn off this hack and make `&&`
+and `||` behave as you would expect.
+
+As a side-effect of these changes, the `assert` and `debug_assert` macros
+become usable in a const context if `#![feature(const_panic)]` is also
+enabled. However, the other assert macros (e.g., `assert_eq`,
+`debug_assert_ne`) remain forbidden, since they need to call `Debug::fmt` on
+their arguments.
+
+Also forbidden are looping constructs, `while`, `for` and `loop`, which will
+be [feature-gated separately][52000], and the `?` operator, which calls
+`From::from` on the value inside the `Err` variant. The design for
+`const` trait methods is still being discussed.
+
+[52000]: https://github.com/rust-lang/rust/issues/52000
+
+## What's next?
+
+This change will allow a great number of standard library functions to be
+made `const`. If you like, you can help with this process! There's a [list of
+numeric functions][const-int] that can be constified with little effort. Be
+aware that things move slowly in the standard library, and it's always best
```

Comment:
Constification PRs have been [closed](https://github.com/rust-lang/rust/pull/65341) because they don't have a motivating use case. I agree with your position on this, but is there consensus? Regardless, "slowly" is unnecessarily pejorative.

Reply:
I believe in the case you mention, it's more the case that we know that it is not useful in const contexts yet rather than requiring an elaborate justification. Here I think "proving out the language feature" is a useful justification in-and-of-itself. The libs team will have an opportunity to reject stabilization and revert things (if e.g. it incurred hacks), so I think we should be good. My interest here is to have a light-weight process and avoid bureaucracy.

Reply:
Okay, I've removed that sentence entirely. Do we want to suggest including a concrete example? For many functions that will be overkill.

Reply:
Maybe we could give an example of how you constify something by pointing to some merged PR which introduced `rustc_const_unstable`?

Reply:
How about rust-lang/rust#63123?

Reply:
Yeah that works; https://github.com/rust-lang/rust/pull/61635 is another good example.

Reply:
I used rust-lang/rust#61635 as the example and explicitly mention `rustc_const_unstable`.